clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3564,23 +3543,188 @@ static void handleSYCLIntelSchedulerTargetFmaxMhzAttr(Sema &S, Decl *D,
 }
 
 // Handles max_global_work_dim.
-static void handleMaxGlobalWorkDimAttr(Sema &S, Decl *D, const ParsedAttr &A) {
-  if (D->isInvalidDecl())
-    return;
+// Returns a ArgCheck value; EqualToOne means all argument values are one,
+// NotEqualToOne means all argument values are not one, and Unknown means that
+// the argument values are invalid.
+enum class ArgCheck { Unknown, EqualToOne, NotEqualToOne };
+static ArgCheck AreArgValuesToOne(const Expr *E, const Expr *E1,
+		                  const Expr *E2) {
+  // If any of the expression is value dependent, test is invalid.
+  const auto *XDimExpr = dyn_cast<ConstantExpr>(E);
+  const auto *YDimExpr = dyn_cast<ConstantExpr>(E1);
+  const auto *ZDimExpr = dyn_cast<ConstantExpr>(E2);
+  if (!XDimExpr || !YDimExpr || !ZDimExpr)
+    return ArgCheck::Unknown;
+
+  // Check the argument values.
+  return (XDimExpr->getResultAsAPSInt() == 1 ||
+          YDimExpr->getResultAsAPSInt() == 1 ||
+          ZDimExpr->getResultAsAPSInt() == 1)
+          ? ArgCheck::EqualToOne
+          : ArgCheck::NotEqualToOne;
+}
+
+void Sema::AddSYCLIntelMaxGlobalWorkDimAttr(Decl *D,
+                                            const AttributeCommonInfo &CI,
+                                            Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
 
-  Expr *E = A.getArgAsExpr(0);
+    // This attribute requires a non-negative value.
+    if (ArgVal < 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*non-negative*/ 1;
+      return;
+    }
```

Comment:
I sort of wonder whether we want to roll this into the [0, 3] range check rather than saying it requires a positive integer (without the range)?